File: scala/higher-order-functions-demo/src/main/scala/FizzBuzzGenerator.scala
import Extensions.Division

class FizzBuzzGenerator() {
    val rules = Array(
        (divisibleBy(15), fixedValue("Fizz-Buzz")),
        (divisibleBy(3), fixedValue("Fizz")),
        (divisibleBy(5), fixedValue("Buzz")),
        (other, returnInput)
    )
    
    def nextNumber(input: Int): String = {
        val rulesThatMatch = rules
            .filter { case (doesTheRuleApply, _) => doesTheRuleApply(input) }
        val transformedResults = rulesThatMatch
            .map { case (_, result) => result(input) }
        transformedResults.head
    }

    private def returnInput = (i: Int) => i.toString

    private def fixedValue(result: String): (Int) => String = (_: Int) => result

    private def other = (_: Int) => true

    private def divisibleBy(divisor: Int) = (i: Int) => i isDivisibleBy divisor
}

File: javascript/fizz-buzz-functional/src/fizz-buzz-generator.js
export let generateFizzBuzz = number => {
    const rules = allRules()
    let result = ''
    for (const r of rules) {
        if (r.appliesTo(number)) {
            if (result) {
                result += '-'
            }
            result += r.result
        }
    }
    return result || number.toString()
}

let allRules = () => [
    {appliesTo: isDivisibleBy(3), result: 'Fizz'},
    {appliesTo: isDivisibleBy(5), result: 'Buzz'},
]

let isDivisibleBy = denominator => number =>
    number % denominator === 0

